exportlegends.lua

```diff
@@ -988,19 +1023,31 @@ function export_detailed_maps()
             local vs = dfhack.gui.getViewscreenByType(df.viewscreen_export_graphical_mapst, 0)
             if not vs then
                 local legends_vs = dfhack.gui.getViewscreenByType(df.viewscreen_legendsst, 0)
-                    or qerror("Could not find legends screen")
+                if not legends_vs then
+                    qerror("Could not find legends screen")
+                end
+
                 gui.simulateInput(legends_vs, 'LEGENDS_EXPORT_DETAILED_MAP')
             end
 
             vs = dfhack.gui.getViewscreenByType(df.viewscreen_export_graphical_mapst, 0)
-                or qerror("Could not find map export screen")
+            if not vs then
+                qerror("Could not find map export screen")
+            end
+            
             vs.sel_type = i - 1
-            print('    Exporting map: ' .. MAPS[i])
+            -- Move into the save folder
+            if not move_to_save_folder() then
+                qerror('Could not move into the save folder.')
+            end
+            print('    Exporting map ' ..i.. '/' ..#MAPS..': '.. MAPS[i])
             gui.simulateInput(vs, 'SELECT')
             while dfhack.gui.getCurViewscreen() == vs do
                 script.sleep(10, 'frames')
             end
+            move_back_to_main_folder() -- Move back out of the save folder
```

Comment:
Another thing that would be useful here, from the [Lua API docs](https://docs.dfhack.org/en/0.47.04-r1/docs/Lua%20API.html#locking-and-finalization), is `dfhack.with_finalizer`. That will allow you to specify a callback and a cleanup function (`move_back_to_main_folder`) that will *always* get called, even if the callback throws an error. I think that would be a good idea here because DF will probably break down the line if the working directory isn't restored properly. (Not that the script is throwing any errors currently, but if it ever does in the future, we could run into issues.)

In this case, you'd probably want to wrap the entire anonymous function passed to script.start:

```lua
script.start(dfhack.with_finalizer(move_back_to_main_folder, function()
  ...
end))
```

(`dfhack.with_finalizer(move_back_to_main_folder, export_detailed_maps)` from the caller would *not* work, since export_detailed_maps is asynchronous)

Reply:
Making it future proof is always a good plan.
Changed it, had to make it a bit bigger, but added comments to make everything clear.